tidy3d/plugins/design/design.py

```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
-            Can be converted to ``pandas.DataFrame`` with ``.to_dataframe()``.
-        """
-
-        # run the functions using the `method.run_batch`
-        fn_args, fn_values, task_ids, batch_data = self.method.run_batch(
-            parameters=self.parameters,
-            fn_pre=fn_pre,
-            fn_post=fn_post,
-            path_dir=path_dir,
-            **batch_kwargs,
-        )
-
-        # store the pre and post functions in a single source code
-        fn_source_pre = self.get_fn_source(fn_pre)
-        fn_source_post = self.get_fn_source(fn_post)
-        fn_source = self._make_batch_fn_source(fn_source_pre, fn_source_post)
-
-        # package the result
+        # Package the result
         return self._package_run_results(
             fn_args=fn_args,
             fn_values=fn_values,
             fn_source=fn_source,
-            task_ids=task_ids,
-            batch_data=batch_data,
+            aux_values=aux_values,
+            task_ids=self._task_names,
         )
+
+    def run_single(self, fn: Callable):
+        """Run a single function of parameter inputs."""
+        evaluate_fn = self._get_evaluate_fn_single(fn=fn)
+        return self.method.run(run_fn=evaluate_fn, parameters=self.parameters)
+
+    def run_pre_post(self, fn_pre: Callable, fn_post: Callable):
+        """Run a function with tidy3d implicitly called in between."""
+        evaluate_fn = self._get_evaluate_fn_pre_post(fn_pre=fn_pre, fn_post=fn_post)
+        return self.method.run(run_fn=evaluate_fn, parameters=self.parameters)
+
+    """ Helpers """
+
+    def _get_evaluate_fn_single(self, fn: Callable) -> list[Any]:
+        """Get function that sequentially evaluates single `fn` for a list of arguments."""
+
+        def evaluate(args_list: list) -> list[Any]:
+            """Evaluate a list of arguments passed to ``fn``."""
+            return [fn(**args) for args in args_list]
+
+        return evaluate
+
+    def _get_evaluate_fn_pre_post(self, fn_pre: Callable, fn_post: Callable) -> list[Any]:
+        """Get function that tries to use batch processing on a set of arguments."""
+
+        def evaluate(args_list: list[tuple[Any, ...]]) -> list[Any]:
+            """Put together into one pipeline."""
+            combined_fn = self._stitch_pre_post(fn_pre=fn_pre, fn_post=fn_post)
+            out = combined_fn(args_list)
+            return out
+
+        return evaluate
+
+    def _stitch_pre_post(self, fn_pre: Callable, fn_post: Callable) -> Callable:
+        """Combine pre and post into one big function, stitching tidy3d calls between if needed."""
+
+        def fn_combined(args_list):
+            sim_dict = {idx: fn_pre(**arg_list) for idx, arg_list in enumerate(args_list)}
+            data = self.fn_mid(sim_dict)
+            post_out = [fn_post(val[1]) for val in data.items()]
+            return post_out
+
+        return fn_combined
+
+    def fn_mid(self, pre_out: Any) -> Any:
```

Comment:
`pre_out` is a `dict[int, Any]` correct? and then this returns something `dict`-like? or more specifically, something that can be iterated over its `.items()`?

Reply:
Yeah you're right. Should the return just be typed as `dict` or is there a clearer way because it can be a BatchData in the case of `dict[int, Simulation]` inputs?

Reply:
I think we can type it as `dict[int, Any]` or we can say `Union[dict[int, Any], BatchData]`